0384-importing-forward-declared-objc-interfaces-and-protocols.md (#1884)
apple/swift-evolution: proposals/NNNN-importing-forward-declared-objc-interfaces-and-protocols.md

# Importing Forward Declared Objective-C Interfaces and Protocols

* Proposal: [SE-NNNN](NNNN-importing-forward-declared-objc-interfaces-and-protocols.md)
* Authors: [Nuri Amari](https://github.com/NuriAmari)
* Review Manager: TBD
* Implementation: https://github.com/apple/swift/pull/61606

## Introduction

This proposal seeks to improve the usability of existing Objective-C libraries from Swift by reducing the negative
impact forward declarations have on API visibility from Swift. We wish to start synthesizing placeholder types to
represent forward declared Objective-C interfaces and protocols in Swift.

Swift-evolution thread: [Discussion thread topic for that proposal](https://forums.swift.org/t/pitch-importing-forward-declared-objective-c-classes-and-protocols/61926)

## Motivation

Forward declarations are very common in many existing Objective-C code bases, used often to break cyclic dependencies or to improve build performance.

> Unfortunately, when it comes to "importability" into Swift they are quite detrimental.
> 
> As it stands, the ClangImporter will fail to import any declaration that references a forward declared type in many common cases. This means a single

Reviewer: I think we can be more concise here and get straight to the point:

```suggestion
Unfortunately, they can cause interoperability problems for Swift. ClangImporter will silently drop declarations that reference a forward declared type in many common cases. This means a single
```

I suggest the wording "silently drop" vs. "fail to import" since "fail" could be interpreted to mean that the compiler fails or emits an error, which isn't the case here; it silently continues and the user is none the wiser.

is empty from the Swift perspective (Swift textual interface generated via swift-ide-test):

_Objective-C_
```objective-c
#import <Foundation/Foundation.h>

@class ForwardDeclaredInterface;
@protocol ForwardDeclaredProtocol;

@interface IncompleteTypeConsumer1 : NSObject
@property id<ForwardDeclaredProtocol> propertyUsingAForwardDeclaredProtocol1;
@property ForwardDeclaredInterface *propertyUsingAForwardDeclaredInterface1;
- (id)init;
- (NSObject<ForwardDeclaredProtocol> *)methodReturningForwardDeclaredProtocol1;
- (ForwardDeclaredInterface *)methodReturningForwardDeclaredInterface1;
- (void)methodTakingAForwardDeclaredProtocol1:
    (id<ForwardDeclaredProtocol>)param;
- (void)methodTakingAForwardDeclaredInterface1:
            (ForwardDeclaredInterface *)param;
@end

ForwardDeclaredInterface *CFunctionReturningAForwardDeclaredInterface1();
void CFunctionTakingAForwardDeclaredInterface1(
    ForwardDeclaredInterface *param);

NSObject<ForwardDeclaredProtocol> *CFunctionReturningAForwardDeclaredProtocol1();
void CFunctionTakingAForwardDeclaredProtocol1(
    id<ForwardDeclaredProtocol> param);
```

_Swift_
```swift
class IncompleteTypeConsumer1 : NSObject {
  init!()
}
```

> We wish to make the experience of consuming an Objective-C API with forward declarations from Swift more consistent with the experience of consuming the API from Objective-C.

Reviewer: You should call out up-front that the user _can_ resolve this today by importing the module providing the full definition of the type, but that the compiler's current behavior doesn't give any feedback that this is what they should do. Emphasize the point here explicitly that this proposal gives the user that important feedback that is currently missing.

Author: Relating to this comment, and the one above about silently dropping declarations: That isn't strictly true, but moreover in my eyes, increased feedback isn't really the goal of this proposal either.

A year or so ago I added diagnostics to the ClangImporter to diagnose these kinds of declarations being dropped. It has been enabled by default for some time: https://forums.swift.org/t/pitch-lazy-clangimporter-diagnostics-enabled-by-default/54651. Notably, these diagnostics are lazy meaning they are only reported if you attempt to use an "unimportable" entity. In that case, you'll get an error like this:

```
/home/nuria/Git/swift-project/swift/scratch/test.h:55:1: warning: method 'methodReturningForwardDeclaredInterface' not imported
- (ForwardDeclaredInterface *) methodReturningForwardDeclaredInterface;
^
/home/nuria/Git/swift-project/swift/scratch/test.h:55:1: note: return type not imported
- (ForwardDeclaredInterface *) methodReturningForwardDeclaredInterface;
^
/home/nuria/Git/swift-project/swift/scratch/test.h:55:1: note: interface 'ForwardDeclaredInterface' is incomplete
- (ForwardDeclaredInterface *) methodReturningForwardDeclaredInterface;
^
/home/nuria/Git/swift-project/swift/scratch/test.h:45:1: note: interface 'ForwardDeclaredInterface' forward declared here
@class ForwardDeclaredInterface;
^
```

If you want eager diagnostics (ie. the diagnostic is not predicated on a usage), I also added that, but it is behind a flag, something like `-enable-eager-clang-importer-diagnostics`, I can't recall the exact spelling. 

If you are suggesting that these diagnostics need to be more explicit about the solution, that's doable, but diagnostics was not really the goal for this proposal. Rather the goal here is to actually import the declaration rather than diagnose the failure.

Reviewer: Ok, that's reasonable. Thanks for the refresher, I wasn't sure what the current state of things was for your previous work. That being the case, some other readers may not be familiar with the history here either, so I think it would be beneficial to link to your previous work and have a brief statement about the diagnostics already shown and how this proposal builds on that by actually synthesizing the type.

class ForwardDeclaredInterface {
}
@available(*, unavailable, message: "This Objective-C protocol has only been forward-declared; import its owning module to use it")
protocol ForwardDeclaredProtocol : NSObjectProtocol {
}
class IncompleteTypeConsumer1 : NSObject {
  var propertyUsingAForwardDeclaredProtocol1: ForwardDeclaredProtocol!
  var propertyUsingAForwardDeclaredInterface1: ForwardDeclaredInterface!
  init!()
  func methodReturningForwardDeclaredProtocol1() -> ForwardDeclaredProtocol!
  func methodReturningForwardDeclaredInterface1() -> ForwardDeclaredInterface!
  func methodTakingAForwardDeclaredProtocol1(_ param: ForwardDeclaredProtocol!)
  func methodTakingAForwardDeclaredInterface1(_ param: ForwardDeclaredInterface!)
}
func CFunctionReturningAForwardDeclaredInterface1() -> ForwardDeclaredInterface!
func CFunctionTakingAForwardDeclaredInterface1(_ param: ForwardDeclaredInterface!)
func CFunctionReturningAForwardDeclaredProtocol1() -> ForwardDeclaredProtocol!
func CFunctionTakingAForwardDeclaredProtocol1(_ param: ForwardDeclaredProtocol!)
```

More usage examples can be found in these tests introduced here: https://github.com/apple/swift/pull/61606

## Detailed design

> Modifications lie almost exclusively in the ClangImporter -- specifically in the SwiftDeclConverter. If asked to convert a

Reviewer: ```suggestion
Modifications lie almost exclusively in ClangImporter -- specifically in `SwiftDeclConverter`. If asked to convert a
```

> `clang::ObjCInterfaceDecl` or `clang::ObjCProtocolDecl` with no definition the SwiftDeclConverter will now return a placeholder type instead of

Reviewer: ```suggestion
`clang::ObjCInterfaceDecl` or `clang::ObjCProtocolDecl` with no definition, `SwiftDeclConverter` will now return a placeholder type instead of
```

bailing. These placeholder types are as described:

```swift
// @class Foo turns into
@available(*, unavailable, message: “This Objective-C class has only been forward declared; import its owning module to use it”)
class Foo : NSObject {}

// @protocol Bar turns into
@available(*, unavailable, message: “This Objective-C protocol has only been forward declared; import its owning module to use it”)
protocol Bar : NSObjectProtocol {}
```

Permitted usages of these types are intentionally limited. You will be able to use Objective-C and C declarations that refer to these types without issue.
You will be able to pass around instances of these incomplete types from Swift to Objective-C and vice versa.

> You’ll also be able to call the set of methods provided by NSObject or NSObjectProtocol on instances.

Reviewer: ```suggestion
You’ll also be able to call the set of methods provided by `NSObject` or `NSObjectProtocol` on instances.
```

My Objective-C is a bit rusty these days; is this also possible on pointers to only-forward-declared class/protocol types today in that language? Does it assume `NSObject(Protocol)` inheritance rather than the possibility of some other root class?

Author: It does make this assumption, but as I understand it Objective-C interop in general already makes this assumption. I'm told that Swift does not know how to interact with objects that don't inherit from NSObject.

I tried a simple example with the release compiler and it crashed at runtime:
```
./src/swift-client-complete
2023-01-04 12:34:03.220 swift-client-complete[43116:24260633] *** NSForwarding: warning: object 0x1041740c0 of class 'Foo' does not implement methodSignatureForSelector: -- trouble ahead
2023-01-04 12:34:03.221 swift-client-complete[43116:24260633] *** NSForwarding: warning: object 0x1041740c0 of class 'Foo' does not implement doesNotRecognizeSelector: -- abort
Trace/BPT trap: 5
```

In general I believe if we are wrong about the assumptions, you'll get a message send failure crash at runtime. I've been assured that NSObject is required for interop, but I'm not sure about `NSObjectProtocol`. It would be helpful to get some feedback from those familiar with the details here. In any case, being able to call the methods provided by `NSObject` and `NSObjectProtocol` is more of a nicety, we can drop that if it turns out to be unsound.

such as declaring a new Swift class that inherits from or conforms to such a type. You will also not be able to create new instances of these types in Swift.

To make these limitations more intuitive to users, a new diagnostic has been added. This diagnostic helps guide users when they attempt to access a member on the incomplete synthesized
type, expecting the complete type. The new diagnostics are shown in the last 4 lines of this example:

```swift
swift-client-fwd-declared.swift:6:7: error: value of type 'Foo' has no member 'sayHello'
myFoo.sayHello()
~~~~~ ^~~~~~~~
<unknown>:0: note: class 'Foo' is a placeholder for a forward declared Objective-C interface and may be missing members; import the definition to access the complete interface
foo-bar-consumer.h:3:1: note: interface 'Foo' forward declared here
@class Foo;
^
```

The feature is gated behind a new frontend flag `-enable-import-objc-forward-declarations`. This flag is on by default for Swift version 6 and onwards. The flag is always disabled in the REPL, as

> allowing it leads to confusing behavior. The [existing ClangImporter diagnostics](https://forums.swift.org/t/pitch-improved-clangimporter-diagnostics/52687) will still

Reviewer: Can you elaborate a bit on what the "confusing behavior" is, so that readers have a better understanding of why it's not supported there?

properly warn users about forward declaration limitations, with or with out the feature enabled.

## Source compatibility

> This change does have the potential to break source compatibility. This is simply a consequence
> of the ClangImporter importing new declarations that were previously hidden. That said,

Reviewer: ```suggestion
This change does have the potential to break source compatibility, since
ClangImporter will now import declarations that were previously dropped. That said,
```

> the impact should rare and easy to fix (just rename or qualify any ambiguous identifiers). 
> 
> As mentioned, to address this, the feature is gated behind a flag for Swift 5 and before, and enabled
> by default in Swift 6.
> 
> That said, the current implementation already passes the source compatibility suite and the source kit stress test.

Reviewer: I think it's fair to say that the impact should be rare, but I would omit the part about it being "easy to fix", since it's much more subjective (a now-colliding decl could somehow be used widely throughout a person's code base, making the refactoring larger).

I would move the sentence about the source compatibility suite and SK stress test up here since. Perhaps something like this for the remainder of the section:

```suggestion
we believe a source compatibility break will be rare; our testing found that our
implementation passes the source compatibility suite and the SourceKit stress test.
Furthermore, the feature is gated behind a flag for Swift 5 and before, so breaking
changes would only occur by enabling the flag or using the new language mode.
```

## Effect on ABI stability

This change has no affect on ABI.

## Effect on API resilience

This change does not influence the ABI characteristics
of public APIs.

## Alternatives considered

The unavailable attribute attached to synthesized declarations might be more restrictive than necessary.
We considered the introduction of a new attribute to denote an incomplete type. This attribute would be used to prevent
operations that are not suitable for an incomplete type (ie. declaring a Swift type that conforms to an incomplete protocol).

Allowing more use of the type however increases the complexity of the problem, and the likelihood of source compatibility issues.
For example, if the synthesized type is not file private, issues could arise with different files having different representations
of the type depending on which modules they import.

> We also attempted to make this implementation usable from the REPL, but similar issues with conflicting "import environments" arose.

Reviewer: Once you elaborate on the specifics of what problems you found in the REPL, I think you can just express this as

```suggestion
We could also make this implementation usable from the REPL, but it would require solving the issues described above.
```

And then add a sentence rationalizing why you don't believe solving those should block this proposal.

of public APIs.

## Alternatives considered

The unavailable attribute attached to synthesized declarations might be more restrictive than necessary.
We considered the introduction of a new attribute to denote an incomplete type. This attribute would be used to prevent
operations that are not suitable for an incomplete type (ie. declaring a Swift type that conforms to an incomplete protocol).

Allowing more use of the type however increases the complexity of the problem, and the likelihood of source compatibility issues.
For example, if the synthesized type is not file private, issues could arise with different files having different representations
of the type depending on which modules they import.

We also attempted to make this implementation usable from the REPL. As mentioned, we had issues with `ClangImporter` caching
imports. We must invalidate the cache, at the very least to allow the new definition of type 'Foo' to be imported. This creates issues
where two incompatible representations of the type (`IncompleteFoo.Foo` and `CompleteFoo.Foo`) are in context at the same time.
To solve these issues we attempted the following:

> 1. Invalidate all direct or indirect references to `IncompleteFoo.Foo` and replace them with `CompleteFoo.Foo`:

Author: Let me know if this is too implementation detail heavy. The TLDR is the REPL problem is technically difficult to crack, more difficult to crack than the non-REPL problem -- at least with my knowledge.
